cpp/src/Ice/Incoming.cpp

```diff
@@ -31,50 +29,42 @@ extern bool printStackTraces;
 
 }
 
-Ice::MarshaledResult::MarshaledResult(const Ice::Current& current) :
-    ostr(make_shared<Ice::OutputStream>(current.adapter->getCommunicator(), Ice::currentProtocolEncoding))
-{
-    ostr->writeBlob(replyHdr, sizeof(replyHdr));
-    ostr->write(current.requestId);
-    ostr->write(replyOK);
-}
-
-IceInternal::IncomingBase::IncomingBase(Instance* instance, ResponseHandler* responseHandler,
-                                        Ice::Connection* connection, const ObjectAdapterPtr& adapter,
-                                        bool response, uint8_t compress, int32_t requestId) :
-    _response(response),
+Incoming::Incoming(Instance* instance, ResponseHandlerPtr responseHandler,
+                                        Ice::ConnectionPtr connection, ObjectAdapterPtr adapter,
+                                        bool twoWay, uint8_t compress, int32_t requestId) :
+    _isTwoWay(twoWay),
     _compress(compress),
     _format(Ice::FormatType::DefaultFormat),
     _os(instance, Ice::currentProtocolEncoding),
-    _responseHandler(responseHandler)
+    _responseHandler(std::move(responseHandler)),
+    _is(nullptr)
 {
-    _current.adapter = adapter;
-    ::Ice::ConnectionI* conn = dynamic_cast<::Ice::ConnectionI*>(connection);
-    _current.con = conn ? conn->shared_from_this() : nullptr;
+    _current.adapter = std::move(adapter);
+    _current.con = std::move(connection);
     _current.requestId = requestId;
     _current.encoding.major = 0;
     _current.encoding.minor = 0;
 }
 
-IceInternal::IncomingBase::IncomingBase(IncomingBase& other) :
-    _current(other._current),
-    _servant(other._servant),
-    _locator(other._locator),
-    _cookie(other._cookie),
-    _response(other._response),
-    _compress(other._compress),
-    _format(other._format),
-    _os(other._os.instance(), Ice::currentProtocolEncoding),
-    _responseHandler(other._responseHandler),
-    _interceptorCBs(other._interceptorCBs)
+Incoming::Incoming(Incoming&& other) :
+    _current(std::move(other._current)),
```

Comment:
Ideally all fields would be movable and we could just use the default move constructor.

Reply:
I would just always create the incoming on the heap. More consistent with the client side and I would no longer worry about such a heap allocation here.

Reply:
It's not just about heap allocation but also about making it a shared object, managed by a shared_ptr.

I find having a stack-allocated Incoming that we move into a heap-allocated shared_ptr<Incoming> for async dispatch (and lambda captures that share the same Incoming) quite idiomatic. But yes, it's more complicated. 

If we embrace heap-allocated object with shared-ownership semantics - useful for async dispatches - then the signature should be:
```cpp
bool _iceDispatch(const IncomingPtr&);
```

> More consistent with the client side 

That's because on the client-side sync invocations are thin wrappers over async invocations. The client-side pipeline is all async. This makes sense since most invocations are async. That's why in IceRPC, InvokeAsync returns a Task, not a ValueTask.

On the other hand, it's common to have truly synchronous dispatches. And that's why with IceRPC, DispatchAsync returns a ValueTask.